README.md

# NinjaDropdown 🐱‍👤

[![style: very good analysis](https://img.shields.io/badge/style-very_good_analysis-B22C89.svg)](https://pub.dev/packages/very_good_analysis)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

NinjaDropdown is a highly customisable dropdown intended for flexibilty while implementing dropdowns.
It enables you to set custom shapes and properties to various dropdown components. Thus enancing look
and feel of complete app.


## Demo 😁

Following are examples of how Ninja Dropdown can be used.
<img src="https://github.com/Swado/ninja_dropdown/raw/master/screenshots/NinjaDropdown.png" alt="drawing" width="700"/>


## Code Snippets 🐱‍💻
* Default

```javascript
NinjaDropdown(
              hintText: "hell0",
              itemList: fruitList,
              onTap: (int val) {},
            ),
```
* Custom trailing icons

```javascript
NinjaDropdown(
              hintText: "hell0",
              itemList: fruitList,
              onTap: (int val) {},
              onOpenTrailingIcon: Icons.grade,
              onCloseTrailingIcon: Icons.cloud,
            ),
```
* Custom Fun

```javascript
NinjaDropdown(
              hintText: "hell0",
              itemList: fruitList,
              onTap: (int val) {},
              onOpenTrailingIcon: Icons.grade,
              onCloseTrailingIcon: Icons.cloud,
              menuBorderRadius: const BorderRadius.only(
                  topLeft: Radius.circular(50),
                  bottomRight: Radius.circular(50)),
              textBoxDecoration: const BoxDecoration(
                  color: Colors.amber,
                  gradient: RadialGradient(
                      colors: [Colors.white, Colors.blue], radius: 5),
                  shape: BoxShape.rectangle,
                  borderRadius: BorderRadius.all(Radius.circular(10))),
              menuColor: Colors.blue[100]!,
              selectedItemColor: Colors.blue[300]!,
            ),
```


## Features ✨

- Customisable textbox
- Customisable Menu card
- Add custom trailing icons
- Custom color properties for elements
- Spacing between Menu card and Textbox.
- Selected item color can be modified.
- Arrow design card can be enabled/disabled.

## Properties 💖

| Property             | Type |Description|                                                               
| ----------------- | ----- | ------------------------------------------------------------------ |
| hintText | String | Hinttext for dropdown. |
| itemList | List<dynamic> | ItemList for populating dropdown. |
| onTap | void Function | Function returns index of selected/tapped item. |
| textBoxDecoration | BoxDecoration? | Property enables customisation of TextBox. |
| isTrailingIcon | bool | Boolean to enable trailing icon. |
| onCloseTrailingIcon | IconData |Set custom trailing icon on close. |
| onOpenTrailingIcon | IconData | Set custom trailing icon on open. |
| spacing | int | Spacing between Textbox and Menu. |
| isArrowEnabled | bool | Boolean to set design element arrow on menu card. |
| menuTextpadding | EdgeInsetsGeometry | Enables padding for menu card items. |
| menuColor | Color | Enables background color for menu card. |
| selectedItemColor | Color | Enables background color for menu card selected item. |
| menuBorderRadius | BorderRadius | Enables border radius for menu. |


## Authors 🙌

- [@Swado](https://github.com/Swado)



CHANGELOG.md
## 0.1.0+1

* First Release for Ninja Dropdown package.

test/ninja_dropdown_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:ninja_dropdown/widget/ninja_dropdown_widget.dart';

void main() {
  final fruitList = [
    'Banana',
    'Apple',
    'kiwi',
    'Orange',
    'Grapes',
    'Custard apple'
  ];
  group('NinjaDropdown tests for text box', () {
    const textboxDecoration = BoxDecoration(
      color: Colors.amber,
      gradient: LinearGradient(colors: [Colors.amber, Colors.red]),
      borderRadius: BorderRadius.only(topLeft: Radius.circular(10)),
    );
    final ninjaPage = MaterialApp(
      home: NinjaDropdown(
        hintText: 'hintText',
        itemList: fruitList,
        onTap: (val) {},
        textBoxDecoration: textboxDecoration,
      ),
    );
    testWidgets('Hint text assigned correctly', (WidgetTester tester) async {
      await tester.pumpWidget(ninjaPage);

      final hintText = find.text('hintText');
      expect(hintText, findsOneWidget);
    });

    testWidgets('textBoxDecoration property check for box radius set correctly',
        (WidgetTester tester) async {
      await tester.pumpWidget(ninjaPage);

      // Check for correct box radius.
      final textBoxCtrl = find.byKey(const Key('textBoxKey'));
      final ctnr = tester.widget<Container>(textBoxCtrl);
      final decoration = ctnr.decoration! as BoxDecoration;
      final borderRadius = decoration.borderRadius! as BorderRadius;

      expect(borderRadius.topLeft, const Radius.circular(10));
    });

    testWidgets('textBoxDecoration property Color and gradient set correctly',
        (WidgetTester tester) async {
      await tester.pumpWidget(ninjaPage);

      // Check for Gradient and color
      final textBoxCtrl = find.byKey(const Key('textBoxKey'));
      final ctnr = tester.widget<Container>(textBoxCtrl);
      final decoration = ctnr.decoration! as BoxDecoration;
      final gradient = decoration.gradient!;
      final color = decoration.color!;

      expect(
        gradient,
        const LinearGradient(colors: [Colors.amber, Colors.red]),
      );
      expect(color, Colors.amber);
    });
  });

  group('Trailing icon tests', () {
    testWidgets('Default trailing icon', (WidgetTester tester) async {
      final ninjaPage = MaterialApp(
        home: NinjaDropdown(
          hintText: 'hintText',
          itemList: fruitList,
          onTap: (val) {},
        ),
      );
      await tester.pumpWidget(ninjaPage);

      final iconCtrl = find.byKey(const Key('onCloseTrailingIcon'));
      final iconCtrl2 = find.byKey(const Key('onOpenTrailingIcon'));

      // Inital closed dropdown state
      expect(iconCtrl, findsOneWidget);
      expect(iconCtrl2, findsNothing);
      await tester.tap(iconCtrl);
      await tester.pump(const Duration(microseconds: 10));

      // Open dropdown state
      expect(iconCtrl, findsNothing);
      expect(iconCtrl2, findsOneWidget);
    });
    testWidgets('isTrailingIcon set to false', (WidgetTester tester) async {
      final ninjaPage = MaterialApp(
        home: NinjaDropdown(
          hintText: 'hintText',
          itemList: fruitList,
          onTap: (val) {},
          isTrailingIcon: false,
        ),
      );
      await tester.pumpWidget(ninjaPage);

      final iconCtrl = find.byKey(const Key('onCloseTrailingIcon'));
      final iconCtrl2 = find.byKey(const Key('onOpenTrailingIcon'));
      final textBoxCtrl = find.byKey(const Key('textBoxKey'));

      // Inital closed dropdown state
      expect(iconCtrl, findsNothing);
      expect(iconCtrl2, findsNothing);

      await tester.tap(textBoxCtrl);
      await tester.pump(const Duration(microseconds: 10));

      // Open dropdown state
      expect(iconCtrl, findsNothing);
      expect(iconCtrl2, findsNothing);
    });
    testWidgets('isTrailingIcon set to true and custom icons passed',
        (WidgetTester tester) async {
      final ninjaPage = MaterialApp(
        home: NinjaDropdown(
          hintText: 'hintText',
          itemList: fruitList,
          onTap: (val) {},
          onCloseTrailingIcon: Icons.cloud,
          onOpenTrailingIcon: Icons.star,
        ),
      );
      await tester.pumpWidget(ninjaPage);

      final iconCtrl = find.byKey(const Key('onCloseTrailingIcon'));
      final iconCtrl2 = find.byKey(const Key('onOpenTrailingIcon'));

      // Trailing Icon for onClose
      final ctrl = tester.widget<Icon>(iconCtrl);
      expect(ctrl.icon, Icons.cloud);
      expect(find.byIcon(Icons.cloud), findsOneWidget);

      await tester.tap(iconCtrl);
      await tester.pump(const Duration(microseconds: 10));

      // Open dropdown state
      final ctrl2 = tester.widget<Icon>(iconCtrl2);
      expect(ctrl2.icon, Icons.star);
      expect(find.byIcon(Icons.star), findsOneWidget);
    });
  });

  group('Menu properties', () {
    final ninjaPage = MaterialApp(
      home: NinjaDropdown(
        hintText: 'hintText',
        itemList: fruitList,
        onTap: (val) {},
        spacing: 30,
        menuColor: Colors.yellow,
        menuBorderRadius:
            const BorderRadius.horizontal(left: Radius.circular(10)),
      ),
    );
    testWidgets('Arrow enabled for menu card', (WidgetTester tester) async {
      await tester.pumpWidget(ninjaPage);
      final arrowKey = find.byKey(const Key('arrowContainerKey'));
      final textBoxCtrl = find.byKey(const Key('textBoxKey'));

      // Open menu card.
      await tester.tap(textBoxCtrl);
      await tester.pump(const Duration(microseconds: 10));

      // Check if arrow present.
      expect(arrowKey, findsOneWidget);
    });
    testWidgets('Arrow disabled for menu card', (WidgetTester tester) async {
      final ninjaPage = MaterialApp(
        home: NinjaDropdown(
          hintText: 'hintText',
          itemList: fruitList,
          onTap: (val) {},
          isArrowEnabled: false,
        ),
      );
      await tester.pumpWidget(ninjaPage);
      final arrowKey = find.byKey(const Key('arrowContainerKey'));
      final textBoxCtrl = find.byKey(const Key('textBoxKey'));

      // Open menu card.
      await tester.tap(textBoxCtrl);
      await tester.pump(const Duration(microseconds: 10));

      // Check if arrow absent.
      expect(arrowKey, findsNothing);
    });
    testWidgets('Menu card and Arrow color', (WidgetTester tester) async {
      await tester.pumpWidget(ninjaPage);
      final textBoxCtrl = find.byKey(const Key('textBoxKey'));

      // Open menu card.
      await tester.tap(textBoxCtrl);
      await tester.pump(const Duration(microseconds: 10));

      final arrowKey = find.byKey(const Key('arrowContainerKey'));
      final arrowWidget = tester.widget<Container>(arrowKey);
      final menuCardKey = find.byKey(const Key('menuCardKey'));
      final menuCardWidget = tester.widget<Material>(menuCardKey);
      // Check for menu card and arrow color.
      expect(menuCardWidget.color, Colors.yellow);
      expect(arrowWidget.color, Colors.yellow);
    });
    testWidgets('Menu card border radius', (WidgetTester tester) async {
      await tester.pumpWidget(ninjaPage);
      final textBoxCtrl = find.byKey(const Key('textBoxKey'));

      // Open menu card.
      await tester.tap(textBoxCtrl);
      await tester.pump(const Duration(microseconds: 10));

      final menuCardKey = find.byKey(const Key('menuCardKey'));
      final menuCardWidget = tester.widget<Material>(menuCardKey);
      // Check for menu card color.
      expect(
        menuCardWidget.shape,
        const RoundedRectangleBorder(
          borderRadius: BorderRadius.horizontal(left: Radius.circular(10)),
        ),
      );
    });
  });
}


example/lib/main.dart
import 'package:example/example1.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a blue toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const Example1(title: 'Ninja Dropdown'),
    );
  }
}


linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "ninja_dropdown");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "ninja_dropdown");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.Create(L"ninja_dropdown", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


lib/widget/ninja_dropdown_widget.dart
import 'package:flutter/material.dart';

/// Customisable dropdown allowing developers with more
/// flexibilty while configuring dropdown.
class NinjaDropdown extends StatefulWidget {
  ///Creates configuration for NinjaDropdown.
  const NinjaDropdown({
    required this.hintText,
    required this.itemList,
    required this.onTap,
    super.key,
    this.textBoxDecoration,
    this.isTrailingIcon = true,
    this.onCloseTrailingIcon = Icons.arrow_drop_down,
    this.onOpenTrailingIcon = Icons.arrow_drop_up,
    this.spacing = 10,
    this.isArrowEnabled = true,
    this.menuTextpadding = const EdgeInsets.all(8),
    this.menuColor = Colors.white,
    this.selectedItemColor = Colors.black12,
    this.menuBorderRadius = const BorderRadius.all(Radius.circular(10)),
  });

  /// Hinttext for dropdown.
  final String hintText;

  /// ItemList for populating dropdown.
  final List<dynamic> itemList;

  /// onTap Function returns index of selected/tapped item.
  final void Function(int index) onTap;

  /// textBoxDecoration Property enables customisation of TextBox.
  /// Supported properties are color,border,shape,borderRadius,gradient.
  final BoxDecoration? textBoxDecoration;

  /// Boolean to enable trailing icon. Default [bool true].
  final bool isTrailingIcon;

  /// Set custom trailing icon on close. Default [Icons.arrow_drop_down].
  final IconData onCloseTrailingIcon;

  /// Set custom trailing icon on open. Default [Icons.arrow_drop_up].
  final IconData onOpenTrailingIcon;

  /// Spacing between Textbox and Menu. Default [10].
  final int spacing;

  /// Boolean to set design element arrow on menu card. Default [bool true].
  final bool isArrowEnabled;

  /// menuTextPadding enables padding for menu card items.
  final EdgeInsetsGeometry menuTextpadding;

  /// menuColor enables background color for menu card. Default [Colors.white].
  final Color menuColor;

  /// selectedItemColor enables background color for menu card selected item.
  /// Default [Colors.black12].
  final Color selectedItemColor;

  /// menuBorderRadius enables border radius for menu.
  /// Default [Radius.circular(10)].
  final BorderRadius menuBorderRadius;

  @override
  State<NinjaDropdown> createState() => _NinjaDropdownState();
}

class _NinjaDropdownState extends State<NinjaDropdown> {
  GlobalKey? actionKey;
  bool isDropdownOpen = false;
  double height = 0;
  double width = 0;
  double xPosition = 0;
  double yPosition = 0;
  OverlayEntry? floatingOverlayMenu;
  int? selectedIndex;
  @override
  void initState() {
    actionKey = LabeledGlobalKey(widget.hintText);
    super.initState();
  }

  void dropdownData() {
    final renderBox =
        actionKey!.currentContext!.findRenderObject()! as RenderBox;
    height = renderBox.size.height;
    width = renderBox.size.width;
    final offset = renderBox.localToGlobal(Offset.zero);
    xPosition = offset.dx;
    yPosition = offset.dy;
  }

  OverlayEntry _createOverlayEntry() {
    return OverlayEntry(
      builder: (context) => Positioned(
        left: xPosition,
        width: width,
        top: yPosition + height + widget.spacing,
        child: SizedBox(
          height: height * 5,
          child: Column(
            children: [
              if (widget.isArrowEnabled)
                Align(
                  child: ClipPath(
                    clipper: ArrowClipper(),
                    child: Material(
                      elevation: 4,
                      color: Colors.black,
                      child: Container(
                        key: const Key('arrowContainerKey'),
                        height: 10,
                        width: 20,
                        color: widget.menuColor,
                      ),
                    ),
                  ),
                ),
              Expanded(
                child: Material(
                  key: const Key('menuCardKey'),
                  color: widget.menuColor,
                  shape: RoundedRectangleBorder(
                    borderRadius: widget.menuBorderRadius,
                  ),
                  clipBehavior: Clip.hardEdge,
                  elevation: 30,
                  child: ListView.builder(
                    physics: const ScrollPhysics(
                      parent: AlwaysScrollableScrollPhysics(),
                    ),
                    itemCount: widget.itemList.length,
                    itemBuilder: (context, index) => GestureDetector(
                      onTap: () {
                        setState(() {
                          selectedIndex = index;
                          floatingOverlayMenu?.remove();
                          floatingOverlayMenu = null;
                          widget.onTap(index);
                          isDropdownOpen = !isDropdownOpen;
                        });
                      },
                      child: Container(
                        padding: widget.menuTextpadding,
                        decoration: BoxDecoration(
                          color: index == selectedIndex
                              ? widget.selectedItemColor
                              : null,
                          border: const Border(
                            bottom: BorderSide(
                              color: Color.fromARGB(
                                255,
                                237,
                                236,
                                236,
                              ),
                            ),
                          ),
                        ),
                        child: Text(
                          widget.itemList[index].toString(),
                          style: const TextStyle(
                            fontSize: 20,
                            color: Colors.black,
                          ),
                        ),
                      ),
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      key: actionKey,
      onTap: () {
        setState(() {
          if (isDropdownOpen) {
            floatingOverlayMenu?.remove();
            floatingOverlayMenu = null;
          } else {
            dropdownData();
            floatingOverlayMenu = _createOverlayEntry();
            Overlay.of(context).insert(floatingOverlayMenu!);
          }
          isDropdownOpen = !isDropdownOpen;
        });
      },
      child: textBox(),
    );
  }

  Container textBox() {
    return Container(
      key: const Key('textBoxKey'),
      width: 200,
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(20),
        color: Colors.white,
        border: Border.all(),
      ).copyWith(
        color: widget.textBoxDecoration?.color,
        border: widget.textBoxDecoration?.border,
        shape: widget.textBoxDecoration?.shape,
        borderRadius: widget.textBoxDecoration?.borderRadius,
        gradient: widget.textBoxDecoration?.gradient,
      ),
      padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 8),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: [
          Flexible(
            child: Text(
              selectedIndex != null
                  ? widget.itemList[selectedIndex!].toString()
                  : widget.hintText,
              overflow: TextOverflow.ellipsis,
            ),
          ),
          if (widget.isTrailingIcon && isDropdownOpen)
            Icon(
              key: const Key('onOpenTrailingIcon'),
              widget.onOpenTrailingIcon,
            )
          else if (widget.isTrailingIcon && !isDropdownOpen)
            Icon(
              key: const Key('onCloseTrailingIcon'),
              widget.onCloseTrailingIcon,
            )
        ],
      ),
    );
  }
}

/// ArrowClipper class for creating custom clippath.
class ArrowClipper extends CustomClipper<Path> {
  @override
  Path getClip(Size size) {
    final path = Path()
      ..moveTo(0, size.height)
      ..lineTo(size.width / 2, 0)
      ..lineTo(size.width, size.height);
    return path;
  }

  @override
  bool shouldReclip(covariant CustomClipper<Path> oldClipper) => true;
}


example/lib/example1.dart
import 'package:flutter/material.dart';
import 'package:ninja_dropdown/ninja_dropdown.dart';

class Example1 extends StatefulWidget {
  final String title;
  const Example1({super.key, required this.title});

  @override
  State<Example1> createState() => _Example1State();
}

class _Example1State extends State<Example1> {
  @override
  Widget build(BuildContext context) {
    List fruitList = [
      "Banana",
      "Apple",
      "kiwi",
      "Orange",
      "Grapes",
      "Custard apple"
    ];
    return Scaffold(
      appBar: AppBar(title: Text(widget.title)),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          Center(
            child: NinjaDropdown(
              hintText: "hell0",
              itemList: fruitList,
              onTap: (int val) {},
              onOpenTrailingIcon: Icons.grade,
              onCloseTrailingIcon: Icons.cloud,
              // menuBorderRadius: const BorderRadius.only(
              //     topLeft: Radius.circular(50),
              //     bottomRight: Radius.circular(50)),
              // textBoxDecoration: const BoxDecoration(
              //     color: Colors.amber,
              //     gradient: RadialGradient(
              //         colors: [Colors.white, Colors.blue], radius: 5),
              //     shape: BoxShape.rectangle,
              //     borderRadius: BorderRadius.all(Radius.circular(10))),
              // menuColor: Colors.blue[100]!,
              // selectedItemColor: Colors.blue[300]!,
            ),
          ),
        ],
      ),
    );
  }
}
